xclim/indices/fire/_ffdi.py

```diff
@@ -0,0 +1,393 @@
+    # a guvectorized function which has the output(s) in its function signature
+    return _keetch_byram_drought_index(pr, tasmax, pr_annual, kbdi0)
+
+
+@declare_units(
+    pr="[precipitation]",
+    tasmax="[temperature]",
+    pr_annual="[precipitation]",
+    kbdi0="[precipitation]",
+)
+def keetch_byram_drought_index(
+    pr: xr.DataArray,
+    tasmax: xr.DataArray,
+    pr_annual: xr.DataArray,
+    kbdi0: xr.DataArray | None = None,
+) -> xr.DataArray:
+    """Calculate the Keetch-Byram drought index (KBDI) for soil moisture deficit.
+
+    This method implements the methodology and formula described in :cite:t:`ffdi-finkele_2006`
+    (section 2.1.1) for calculating the KBDI. See Notes below.
+
+    Parameters
+    ----------
+    pr : xr.DataArray
+        Total rainfall over previous 24 hours [mm/day].
+    tasmax : xr.DataArray
+        Maximum temperature near the surface over previous 24 hours [degC].
+    pr_annual: xr.DataArray
+        Mean (over years) annual accumulated rainfall [mm/year].
+    kbdi0 : xr.DataArray, optional
+        Previous KBDI map used to initialise the KBDI calculation [mm/day]. Defaults to 0.
+
+    Returns
+    -------
+    kbdi : xr.DataArray
+        Keetch-Byram drought index.
+
+    Notes
+    -----
+    :cite:t:`ffdi-finkele_2006` limits the maximum KBDI to 200 mm to represent the maximum field capacity of the soil
+    (8 in according to :cite:t:`ffdi-keetch_1968`). However, it is more common in the literature to limit the KBDI to
+    203.2 mm which is a more accurate conversion from in to mm. In the function, the KBDI is limited to 203.2 mm.
+
+    References
+    ----------
+    :cite:cts:`ffdi-keetch_1968,ffdi-finkele_2006,ffdi-holgate_2017,ffdi-dolling_2005`
+    """
+    pr = convert_units_to(pr, "mm/day")
+    tasmax = convert_units_to(tasmax, "C")
+    pr_annual = convert_units_to(pr_annual, "mm/year")
+    if kbdi0 is not None:
+        kbdi0 = convert_units_to(kbdi0, "mm/day")
+    else:
+        kbdi0 = xr.full_like(pr.isel(time=0), 0)
+
+    kbdi = xr.apply_ufunc(
+        _keetch_byram_drought_index_pass,
+        pr,
+        tasmax,
+        pr_annual,
+        kbdi0,
+        input_core_dims=[["time"], ["time"], [], []],
+        output_core_dims=[["time"]],
+        dask="parallelized",
+        output_dtypes=[pr.dtype],
+    )
+    kbdi.attrs["units"] = "mm/day"
+    return kbdi
+
+
+def _griffiths_drought_factor_pass(pr, smd, lim):
+    """Pass inputs on to guvectorized function `_griffiths_drought_factor`. DO NOT CALL DIRECTLY, use `griffiths_drought_factor` instead."""
+    # This function is actually only required as xr.apply_ufunc will not receive
+    # a guvectorized function which has the output(s) in its function signature
+    return _griffiths_drought_factor(pr, smd, lim)
+
+
+@declare_units(
+    pr="[precipitation]",
+    smd="[precipitation]",
+)
+def griffiths_drought_factor(
+    pr: xr.DataArray,
+    smd: xr.DataArray,
```

Comment:
Please add an entry to `xclim/data/variables.yml` for this new variable.

Reply:
I note that "soil moisture deficit" doesn't seem to exist in the CMIP6 models and thus there is no official variable name to follow. From my quick search, `smd` seems the most commonly used, even though CMIP6 uses `mrso` for soil moisture.

Reply:
Added. I went through the same process as you looking for a CMIP name and found nothing. In `xlim/data/variables.yml` I've given `smd` canonical units of mm since that seemed like the most obvious choice.